reactor-extra/src/main/java/reactor/scheduler/clock/SchedulerClock.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2011-2017 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.scheduler.clock;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.concurrent.TimeUnit;
+
+import reactor.core.scheduler.Scheduler;
+import reactor.util.annotation.NonNull;
+
+public class SchedulerClock extends Clock {
+
+	private final Scheduler scheduler;
+	private final ZoneId    zone;
+
+	private SchedulerClock(Scheduler scheduler, ZoneId zone) {
+		this.scheduler = scheduler;
+		this.zone = zone;
+	}
+
+	@Override
+	public ZoneId getZone() {
+		return zone;
+	}
+
+	@Override
+	public SchedulerClock withZone(ZoneId zone) {
+		return new SchedulerClock(scheduler, zone);
+	}
+
+	public Scheduler getScheduler() {
+		return scheduler;
+	}
+
+	public SchedulerClock withScheduler(Scheduler scheduler) {
+		return new SchedulerClock(scheduler, zone);
+	}
+
+	@Override
+	public long millis() {
+		return scheduler.now(TimeUnit.MILLISECONDS);
+	}
+
+	@Override
+	public Instant instant() {
+		return Instant.ofEpochMilli(millis());
```

Comment:
for future compatibility, this might want to use `Instant.ofEpochSecond(epochSecond, nanoAdjustment`. in Java 9 on Linux `Instant.now()` from the system `Clock` instances has microsecond granularity

Reply:
Actually, I see no sense in using that API, because under the hood both methods work similarly 
``` Java
	public static Instant ofEpochSecond(long epochSecond, long nanoAdjustment) {
        long secs = Math.addExact(epochSecond, Math.floorDiv(nanoAdjustment, NANOS_PER_SECOND));
        int nos = (int)Math.floorMod(nanoAdjustment, NANOS_PER_SECOND);
        return create(secs, nos);
    }

    public static Instant ofEpochMilli(long epochMilli) {
        long secs = Math.floorDiv(epochMilli, 1000);
        int mos = (int)Math.floorMod(epochMilli, 1000);
        return create(secs, mos * 1000_000);
    }
```

Reply:
they work similarly, but `ofEpochMilli` is limited to millisecond-resolution.

Reply:
Still do not get it... to extract millis from `Scheduler#now` will be needed the similar math as in `ofEpochMilli`

Reply:
ah i meant to ask the `Scheduler` for `now` in microseconds, and to use that.

Reply:
so future proof as in "if at some point the Scheduler becomes capable of microsecond resolution"?

Reply:
I believe that we may postpone it to that changes (I mean change current implementation or even add it to Reactor later after incubation)